Bluetooth: Controller: nRF54Lx: Use NRF_GRTC for radio scheduling

cvinayak opened the pull request:
Use NRF_GRTC, Global real-time counter, instead of NRF_RTC.

NRF_GRTC is present in a different power domain in the SoC.
Also, NRF_GRTC provide microsecond resolution timing units,
in the future use of NRF_TIMER for packet timer can also be
replaced.

Use of NRF_RTC would keep Radio power domain ON in addition
to already ON NRF_GRTC's power domain, hence switch to using
NRF_GRTC on nRF54Lx SoC to have lower power consumptions.

Add ticker implementation support for free running counter use.

cvinayak committed:
samples: Bluetooth: hci_uart: Add preliminary DF support for nrf54l15pdk

Add preliminary direction finding support building hci_uart
sample for nrf54l15pdk.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Review rework GRTC support

Review rework GRTC support for nRF54Lx SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use NRF_GRTC for radio scheduling

Use NRF_GRTC, Global real-time counter, instead of NRF_RTC.

NRF_GRTC is present in a different power domain in the SoC.
Also, NRF_GRTC provide microsecond resolution timing units,
in the future use of NRF_TIMER for packet timer can also be
replaced.

Use of NRF_RTC would keep Radio power domain ON in addition
to already ON NRF_GRTC's power domain, hence switch to using
NRF_GRTC on nRF54Lx SoC to have lower power consumptions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Support Radio fast ramp up

Add support for Radio fast ramp up for nRF54Lx SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Remove redundant HAL_TICKER_CNTR_CLK_FREQ_HZ

Remove redundant HAL_TICKER_CNTR_CLK_FREQ_HZ define.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add HAL_TICKER_TICKS_TO_US_64BIT define

Add HAL_TICKER_TICKS_TO_US_64BIT define, to return
microsecond value in 64-bits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Ticker support for free running counter

Add ticker implementation support for free running counter
use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_rx.c:
```diff
@@ -925,14 +925,18 @@ ZTEST(test_rx_unframed, test_rx_time_wrapping)
 	/* Maximum negative difference from 0 */
 	time_now = 0;
-	time_diff = (time_wrapping_point == UINT32_MAX ? INT32_MIN : -ISOAL_TIME_WRAPPING_POINT_US);
+	time_diff = (time_wrapping_point == UINT32_MAX) ?
+		    INT32_MIN :
+		    -((int64_t)ISOAL_TIME_WRAPPING_POINT_US);
```
`time_diff` is `int32_t`, so it doesn't make sense to cast to `int64_t`

niag-Oticon replied:
I think it looks alright. With this timer I would expect `time_wrapping_point == UINT32_MAX` to be true so the value actually assigned should be `INT32_MIN`.

Thalley replied:
> Compiler did not like when ISOAL_TIME_WRAPPING_POINT_US is 32-bit unsigned int, as signed values need an extra bit to represent them in 2's compliment, hence have to typecast to 64-bit, then the compiler is ok to assign it to int32_t.

What is the compiler error/warning? 

Depending on what parameters is provided to e.g. GCC, it would also give an error/warning if you assigned a `int64_t` to a `int32_t`, so I would suggest to do it properly than just replacing one compiler error/warning with a future error/warning. 

One easy solution would be to simply have `time_diff` as `int64_t` and then if the result actually fits in a `int32_t`, then we can cast the result.

cvinayak replied:
The compile error was:

> /home/jack/workspace/zephyrproject/zephyr/tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_rx.c: In function 'test_rx_unframed_test_rx_time_wrapping':
/home/jack/workspace/zephyrproject/zephyr/tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_rx.c:928:21: warning: overflow in conversion from 'long long unsigned int' to 'int32_t' {aka 'int'} changes value from '18446744073197551647' to '-511999969' [-Woverflow]
  928 |         time_diff = (time_wrapping_point == UINT32_MAX ? INT32_MIN : -ISOAL_TIME_WRAPPING_POINT_US);
      |                     ^
/home/jack/workspace/zephyrproject/zephyr/tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_rx.c:935:21: warning: overflow in conversion from 'long long unsigned int' to 'int32_t' {aka 'int'} changes value from '18446744073197551647' to '-511999969' [-Woverflow]
  935 |         time_diff = (time_wrapping_point == UINT32_MAX ? INT32_MIN : -ISOAL_TIME_WRAPPING_POINT_US);
      |                     ^

Thalley commented on tests/bluetooth/ctrl_isoal/src/sub_sets/isoal_test_rx.c:
```diff
@@ -925,14 +925,18 @@ ZTEST(test_rx_unframed, test_rx_time_wrapping)
 	/* Maximum negative difference from maximum time */
 	time_now = ISOAL_TIME_WRAPPING_POINT_US;
-	time_diff = (time_wrapping_point == UINT32_MAX ? INT32_MIN : -ISOAL_TIME_WRAPPING_POINT_US);
+	time_diff = (time_wrapping_point == UINT32_MAX) ?
+		    INT32_MIN :
+		    -((int64_t)ISOAL_TIME_WRAPPING_POINT_US);
```
Ditto

niag-Oticon replied:
Same as above.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -1052,6 +1052,10 @@ config BT_MAYFLY_YIELD_AFTER_CALL
 	  If set to 'n', all pending mayflies for callee are executed before
 	  yielding
 
+config BT_TICKER_CNTR_FREERUNNING
```
Not sure what it is about the name, but `FREERUNNING` hurts my eyes :D Suggest `FREE_RUNNING`

Thalley commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -3157,8 +3173,24 @@ void ticker_job(void *param)
+
+#if defined(CONFIG_BT_TICKER_CNTR_FREERUNNING)
+	if (ticker_id_old_head == TICKER_NULL) {
+		/* No tickers active, synchronize to the free running counter so
+		 * that any new ticker started can have its ticks_to_expire
+		 * relative to current free running counter value.
+		 *
+		 * Note, this if clause is an overhead wherein the check is
+		 * performed for every ticker_job() iteration!
+		 */
+		ticks_previous = ticks_now;
+		instance->ticks_current = ticks_now;
```
Can you elaborate on why `current == previous` here?

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
Commenting here since I can't comment on commit messages: 

The commit message should include

- What is GRTC (it may be a common abbreviation for some, but I have no idea what it is, nor is it ever spelled out in this commit)
- What does it do
- Why are we adding it / what does it solve

cvinayak replied:
updated commit title and message

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -824,6 +824,38 @@ config BT_CTLR_RX_PDU_META
+	help
+	  Enable explicit start of nRF54Lx NRF_GRTC peripheral.
+
+config BT_CTLR_NRF_GRTC_KEEPRUNNING
```
BT_CTLR_NRF_GRTC_KEEP_RUNNING

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -824,6 +824,38 @@ config BT_CTLR_RX_PDU_META
+	help
+	  Keep running nRF GRTC peripheral.
+
+config BT_CTLR_NRF_GRTC_AUTOEN_CPUACTIVE
```
BT_CTLR_NRF_GRTC_AUTO_EN_CPU_ACTIVE

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -824,6 +824,38 @@ config BT_CTLR_RX_PDU_META
+	help
+	  Auto enable nRF GRTC on CPU active.
+
+config BT_CTLR_NRF_GRTC_AUTOEN_DEFAULT
```
`BT_CTLR_NRF_GRTC_AUTO_EN_DEFAULT`

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
There seems to be much less common code than uncommon code in this file. Consider splitting the file into one for GRTC and one without

cvinayak replied:
i believe this will be the first iteration of many to come, but will refactor out to separate file in a future PR once bsim port is functional, safer/easier to refactor with CI coverage in place.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1339,81 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
 	nrf_timer_cc_set(EVENT_TIMER, 0, remainder);
 
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	uint32_t l, h, ho, stale;
```
Are one-letter variables common this these type of files?

What about `l` => `cntr_l`? 
Why is the counter for `H` called both `h` and `ho`?

cvinayak replied:
updated

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1472,7 +1631,13 @@ uint32_t radio_tmr_start_now(uint8_t trx)
 uint32_t radio_tmr_start_get(void)
 {
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	return (uint32_t)(nrf_grtc_sys_counter_cc_get(NRF_GRTC,
+						      GRTC_CC_IDX_RADIO) &
+			  0xffffffff);
```
What is the point of both casting it to `uint32_t` and doing `& 0xffffffff`?

Thalley replied:
> Masking ensures a 32-bit and typecast to make compiler happy.

My point is that casting also ensures this :)

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ticker.c:
```diff
@@ -194,5 +194,9 @@ void hal_ticker_instance0_sched(uint8_t caller_id, uint8_t callee_id, uint8_t ch
 void hal_ticker_instance0_trigger_set(uint32_t value)
 {
-	cntr_cmp_set(0, value);
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	cntr_cmp_set(10U, value);
+#else /* !CONFIG_BT_CTLR_NRF_GRTC */
+	cntr_cmp_set(0U, value);
+#endif /* !CONFIG_BT_CTLR_NRF_GRTC */
```
Magic number

cvinayak replied:
updated

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -120,8 +118,13 @@ static void rtc0_nrf5_isr(const void *arg)
 
 	/* On compare0 run ticker worker instance0 */
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	if (NRF_GRTC->EVENTS_COMPARE[10]) {
```
Magic number

cvinayak replied:
uodated

Thalley commented on samples/bluetooth/hci_uart/boards/nrf54l15pdk_nrf54l15_cpuapp_df.overlay:
```diff
@@ -0,0 +1,33 @@
+
+ &uart20 {
+	compatible = "nordic,nrf-uarte";
+	current-speed = <1000000>;
```
Why this speed over `115200`?

aescolar commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -824,9 +824,42 @@ config BT_CTLR_RX_PDU_META
 	prompt "RX pdu meta data"
 	bool
 
+config BT_CTLR_NRF_GRTC
```
Wouldn't it be better to have this type of settings in a new LLL specific kconfig file? (also the other nordic specific ones which are now in this file)

cvinayak replied:
yes, a future PR to move nRF specific to `Kconfig.ll_sw_split.nrf`, `Kconfig.ll_sw_split.openisa` etc...

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -1052,6 +1086,10 @@ config BT_MAYFLY_YIELD_AFTER_CALL
 	  If set to 'n', all pending mayflies for callee are executed before
 	  yielding
 
+config BT_TICKER_CNTR_FREE_RUNNING
+	# Hidden options to use free running counter in ticker implementation
+	bool
```
Did you consider the approach of using `def_bool` instead of `select`?

cvinayak replied:
no, not aware of it, but will give a try in the future iterative PR with bsim port/coverage is functional.

Thalley replied:
It's basically just `default`ing the value of `BT_TICKER_CNTR_FREE_RUNNING` rather than `select`ing it. 

It's not necessarily better, but worth a consideration

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -11,21 +11,73 @@
+	NRF_GRTC->TASKS_CLEAR = 1U;
+
+#if defined(CONFIG_BT_CTLR_NRF_GRTC_KEEPRUNNING)
+	NRF_GRTC->KEEPRUNNING =
+		(GRTC_KEEPRUNNING_REQUEST1_Active <<
+		 GRTC_KEEPRUNNING_REQUEST1_Pos) &
+		GRTC_KEEPRUNNING_REQUEST1_Msk;
+	NRF_GRTC->TIMEOUT = 0U;
+	NRF_GRTC->INTERVAL = 0U;
+	NRF_GRTC->WAKETIME = 4U;
+#endif /* CONFIG_BT_CTLR_NRF_GRTC_KEEPRUNNING */
+
+	NRF_GRTC->CLKCFG = ((GRTC_CLKCFG_CLKSEL_LFXO <<
+			     GRTC_CLKCFG_CLKSEL_Pos) &
+			    GRTC_CLKCFG_CLKSEL_Msk) |
+			   ((GRTC_CLKCFG_CLKFASTDIV_Min <<
+			     GRTC_CLKCFG_CLKFASTDIV_Pos) &
+			    GRTC_CLKCFG_CLKFASTDIV_Msk);
+#endif /* CONFIG_BT_CTLR_NRF_GRTC_START */
+
+	NRF_GRTC->EVENTS_COMPARE[10] = 0U;
+	NRF_GRTC->INTENSET1 = GRTC_INTENSET1_COMPARE10_Msk;
+
+#if defined(CONFIG_BT_CTLR_NRF_GRTC_START)
+	NRF_GRTC->MODE = ((GRTC_MODE_SYSCOUNTEREN_Enabled <<
+			   GRTC_MODE_SYSCOUNTEREN_Pos) &
+			  GRTC_MODE_SYSCOUNTEREN_Msk) |
+#if defined(CONFIG_BT_CTLR_NRF_GRTC_AUTOEN_CPUACTIVE)
+			 ((GRTC_MODE_AUTOEN_CpuActive <<
+			   GRTC_MODE_AUTOEN_Pos) &
+			  GRTC_MODE_AUTOEN_Msk) |
+#endif /* CONFIG_BT_CTLR_NRF_GRTC_AUTOEN_CPUACTIVE */
+#if defined(CONFIG_BT_CTLR_NRF_GRTC_AUTOEN_DEFAULT)
+			 ((GRTC_MODE_AUTOEN_Default <<
+			   GRTC_MODE_AUTOEN_Pos) &
+			  GRTC_MODE_AUTOEN_Msk) |
+#endif /* CONFIG_BT_CTLR_NRF_GRTC_AUTOEN_DEFAULT */
+			 0U;
+
+	NRF_GRTC->TASKS_START = 1U;
```
Also a couple of magic numbers here. The naming of the fields also makes it hard to determine whether `1` means a numeric value, or a boolean.

cvinayak replied:
some should be replaced now.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -34,7 +86,11 @@ uint32_t cntr_start(void)
 		return 1;
 	}
 
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	/* TODO: */
+#else /* !CONFIG_BT_CTLR_NRF_GRTC */
```
TODO what?

cvinayak replied:
updated

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -11,21 +11,73 @@
+			  GRTC_MODE_SYSCOUNTEREN_Pos) &
+			 GRTC_MODE_SYSCOUNTEREN_Msk;
+
+	NRF_GRTC->TASKS_CLEAR = 1U;
```
You need to use the HAL for all tasks, clearing events, accessing CCL/H, CCEN and anything else w sideeffects.
As is now this just hangs in simulation.

cvinayak replied:
Do you have a list of NRF_GRTC registers with required sideeffects?

aescolar replied:
https://github.com/BabbleSim/ext_nRF_hw_models/blob/main/src/HW_models/NHW_GRTC.h#L19-L40

cvinayak replied:
This is cool, perfect

cvinayak replied:
@aescolar correct me, i could not find bsim hal interfaces for dppi publish for grtc, ppib etc... I have reverted to SOC_SERIES_NRF54LX for NRF_GRTC use. Existing nrf54lxbsim will use NRF_RTC  instead.

aescolar replied:
The files I pointed at contain the registers for which you need to ensure you use the nrf hal. For the others you can use the hal, or if you fancy just do direct register access.
In any case, you still have the *whole* nrf hal interface from the hal headers which will work fine in simulation.

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use nrfx interface for bsim use (3/4)

Use nrfx interface for register access with sideeffects to
facilitate bsim use.

4 part commits:
1. Add nrf_grtc interface.
2. Remove CMSIS interface use.
3. Add nrf_grtc interface, once missing bsim port available.
4. Remove CMSIS interface use, replaced by bsim port.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use nrfx interface for bsim use (1/4)

Use nrfx interface for register access with sideeffects to
facilitate bsim use.

4 part commits:
1. Add nrf_grtc interface.
2. Remove CMSIS interface use.
3. Add nrf_grtc interface, once missing bsim port available.
4. Remove CMSIS interface use, replaced by bsim port.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use nrfx interface for bsim use (4/4)

Use nrfx interface for register access with sideeffects to
facilitate bsim use.

Remove CMSIS interface use, was replaced with nrf_grtc,
nrf_dppic and nrf_ppib interface in previous commit.

4 part commits:
1. Add nrf_grtc interface.
2. Remove CMSIS interface use.
3. Add nrf_grtc interface, once missing bsim port available.
4. Remove CMSIS interface use, replaced by bsim port.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use nrfx interface for bsim use (2/4)

Use nrfx interface for register access with sideeffects to
facilitate bsim use.

Remove CMSIS interface use, was replaced with nrf_grtc
interface in previous commit.

4 part commits:
1. Add nrf_grtc interface.
2. Remove CMSIS interface use.
3. Add nrf_grtc interface, once missing bsim port available.
4. Remove CMSIS interface use, replaced by bsim port.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@carlescufi @aescolar @Thalley if the CI is green, i wish to have the nRF54Lx support (without crypto) in Zephyr v3.7, if possible.

Thalley reviewed (dismissed):
My outstanding comments should not be considered blocking. 

w.r.t merging this for 3.7, it does seems to be a bit too big of a PR to consider for merging now, but leave that up to the release team.

aescolar commented:
> i wish to have the nRF54Lx support (without crypto) in Zephyr v3.7, if possible.

We are way past freeze. Any exception needs to be brought to the release team and get an approving vote by the TSC.
Given that nrf54 support in the controller was added just 1 week ago and as experimental I would say it is not justified to make an exception.

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

cvinayak commented:
Bsim support for many essential peripherals is not implemented for nrf54l15bsim like some GRTC, DPPI, PPIB etc registers required. 

Target PR should not be blocked due to incomplete bsim support.

cvinayak commented:
> Bsim support for many essential peripherals is not implemented for nrf54l15bsim like some GRTC, DPPI, PPIB etc registers required.
> 
> Target PR should not be blocked due to incomplete bsim support.

Refer to: https://github.com/zephyrproject-rtos/zephyr/pull/74183#discussion_r1640773507

@aescolar you have blocked this PR for reasons related to simulation support (https://github.com/zephyrproject-rtos/zephyr/pull/74183#discussion_r1639692684), but please compare for example: https://github.com/zephyrproject-rtos/nrf_hw_models/blob/6c389b9b5fa0a079cd4502e69d375da4c0c289b7/src/nrfx/hal/nrf_grtc.c  it is missing nrfx hal interfaces (like publish_set).
Same with implementation of PPIB https://github.com/zephyrproject-rtos/nrf_hw_models/blob/6c389b9b5fa0a079cd4502e69d375da4c0c289b7/src/nrfx/hal/nrf_grtc.c missing publish interfaces. bsim may have preliminary support, but this should not block target implementation using CMSIS/nrfx hal and driver interface

aescolar commented:
>  but please compare for example:... it is missing nrfx hal interfaces (like publish_set). Same with implementation of PPIB ..

Those files you pointed at only contain the functions which need to be overriden/replaced compared to the nrf hal files.
The whole nrf hal is still included. The functions to configure the publish registers come from the plain nrf hal.

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -5,27 +5,92 @@
+	 *        with ability to select/set IRQ group.
+	 *        Shared interrupts is an option? It may add ISR latencies?
+	 */
+	NRF_GRTC->INTENSET1 = HAL_CNTR_GRTC_INTENSET_COMPARE_TICKER_Msk;
```
How come you want to explicitly use this interrupt instead of whichever the HAL has allocated for this processor and secure/nonsecure mode?
(note that the accesses thru the hal to SYSCOUNTER will be thru the "domain" the HAL allocated for this processor)

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
+		cntr_h = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+		cntr_l = nrf_grtc_sys_counter_low_get(NRF_GRTC);
+		cntr_h_overflow = nrf_grtc_sys_counter_high_get(NRF_GRTC);
+	} while ((cntr_h & GRTC_SYSCOUNTER_SYSCOUNTERH_BUSY_Msk) ||
```
This logic seems incorrect.
You should be reading the low part first and then the high.
If the high does not have the overflow bit set, the whole 52 bit value is correct.

(Reading the high first is against the spec, and I presume you will be very likely to read an overflow from that first high read, as that would be relative to the last time the low was read)

aescolar replied:
`nrfy_grtc_sys_counter_get()`

aescolar replied:
Discussed offline, rewriting this in a more optimal way can be done later

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	uint32_t cntr_l, cntr_h, cntr_h_overflow, stale;
+
+	/* NOTE: We are going to use TASKS_CAPTURE to read current
```
I don't follow why you are doing this.
You should be able to just read SYSCOUNTER to get the current GRTC counter value.
Why do you use the capture task instead?

aescolar replied:
I just meant that it seems the use of the task to capture could have been instead replaced by using nrfy_grtc_sys_counter_get()

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
+
+	/* Read current syscounter value */
+	do {
+		cntr_h = nrf_grtc_sys_counter_high_get(NRF_GRTC);
```
Same about high low high read

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
+				    ((((uint64_t)cntr_h & GRTC_CC_CCH_CCH_Msk) << 32) | value));
+
+	/* Enable compare */
+	nrf_grtc_sys_counter_compare_event_enable(NRF_GRTC, cmp);
```
This would be redundant.
The CC channel should have been already enabled when doing nrf_grtc_sys_counter_cc_set() (as a sideeffect of writing to CCH )

aescolar replied:
My understanding from the docs is that CC[cc_channel].CCEN is automatically set when we write to CC[cc_channel].CCH

aescolar replied:
Discussed offline that cleaning this up can be done in a follow up

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+	NRF_DPPIC20->CHENSET = BIT(HAL_EVENT_TIMER_START_PPI);
+
+	/* Setup PPIB send subscribe */
+	NRF_PPIB21->SUBSCRIBE_SEND[HAL_EVENT_TIMER_START_PPI] =
```
Please use `nrf_ppib_subscribe_set()`

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+		 GRTC_PUBLISH_COMPARE_EN_Msk);
+
+	/* Enable same DPPI in Global domain */
+	NRF_DPPIC20->CHENSET = BIT(HAL_EVENT_TIMER_START_PPI);
```
Please use `nrf_dppi_channels_enable()`

aescolar replied:
@cvinayak these 2 commits may help https://github.com/aescolar/zephyr/commits/pr_74183_fix2/

aescolar replied:
Note the PPIB needs to get an index in its PUBLISH_RECEIVE and SUBSCRIBE_SEND registers, not a bitmask

aescolar replied:
With those changes multiple passes in simulation, but you get a pile of warnings due to getting TXEN events in the RADIO when they don't help.
Anyhow, my other comments still apply.

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+
+	/* Read current syscounter value */
+	do {
+		cntr_h = nrf_grtc_sys_counter_high_get(NRF_GRTC);
```
same

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+	 */
+
+	/* Read current syscounter value */
+	do {
```
You are doing this in quite a few places. You may want to consider using `nrfy_grtc_sys_counter_get()`

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+				     ticks_start));
+
+	/* Enable compare */
+	nrf_grtc_sys_counter_compare_event_enable(NRF_GRTC, HAL_CNTR_GRTC_CC_IDX_RADIO);
```
same

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1326,8 +1344,83 @@ uint32_t radio_tmr_start(uint8_t trx, uint32_t ticks_start, uint32_t remainder)
+	NRF_PPIB11->PUBLISH_RECEIVE[HAL_EVENT_TIMER_START_PPI] =
+		BIT(HAL_EVENT_TIMER_START_PPI) | PPIB_PUBLISH_RECEIVE_EN_Msk;
+
+	/* NOTE: We are going to use TASKS_CAPTURE to read current
```
(same)
Anyhow, shouldn't this logic should be refactored into a common function?

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1377,8 +1470,83 @@ uint32_t radio_tmr_start_tick(uint8_t trx, uint32_t tick)
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	uint32_t cntr_l, cntr_h, cntr_h_overflow, stale;
+
+	/* Disable capture/compare */
```
* Same comments as before for the equivalent chunk in `radio_tmr_start()`

* Given that it is so similar to the code in `radio_tmr_start()` should all this not be refactored into a common function?

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi.h:
```diff
@@ -100,7 +100,11 @@ static inline void hal_radio_end_time_capture_ppi_config(void)
 static inline void hal_event_timer_start_ppi_config(void)
 {
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+#else /* !CONFIG_BT_CTLR_NRF_GRTC */
```
Was something forgotten here, or would it make sense for this to just be 
#if !defined(CONFIG_BT_CTLR_NRF_GRTC) ?

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/swi.h:
```diff
@@ -45,7 +45,14 @@
 #define HAL_SWI_RADIO_IRQ  SWI02_IRQn
+
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+#define HAL_SWI_WORKER_IRQ GRTC_1_IRQn
```
Same comment about just using the hal chosen `GRTC_IRQn`

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Use SW_SWITCH_SINGLE_TIMER

Use SW_SWITCH_SINGLE_TIMER for nRF54Lx so that fast ramp
can be used, and hence support assymmetric PHYs in ACL
connections.

nRF54Lx radio domain does not have second timer instance
to efficiently implement software switch unless a second
time is used from a different power domain needed use of
PPIB for the double buffered switch timer compares.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nrf54L15bsim: Use NRF_GRTC

Use NRF_GRTC in nrf54l15bsim board builds.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Revert "tests/bsim: Provisionally disable the nrf54l15 BT tests"

This reverts commit 3d729986e062f9b64f4b888a31e97f7012cc28e0.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor sw_switch hal interface use

Refactor sw_switch hal interface use and document why some
PPI/DPPI channel group related subscriptions are not
disabled explicitly.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ext conn create when using single timer

Fix Extended Connection Creation when using single timer
feature in the Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix sw switch single timer for spurious TXEN/RXEN

Fix software tIFS switching using single timer from
triggering spurious TXEN/RXEN if the first remainder is
shorter than the tIFS delay calculated in the sw_switch()
function.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix BT_CTLR_EARLY_ABORT_PREVIOUS_PREPARE depends

Fix BT_CTLR_EARLY_ABORT_PREVIOUS_PREPARE Kconfig depends on.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: central_hr_peripheral_hr for nrf54l15bsim

Enable central HR peripheral HR samples for nrf54l15bsim.
RealEncryption is BabbleSIM is not enabled as nRF54L port
in Zephyr Controller is not yet supported and the test is
also not using security (SMP pairing).

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF54Lx: Port for data whitening register use

nRF54Lx have updated Data Whitening register settings, add
implementation to correctly set them up.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cntr.c:
```diff
@@ -47,17 +116,89 @@ uint32_t cntr_stop(void)
 	}
 
+#if defined(CONFIG_BT_CTLR_NRF_GRTC)
+	/* TODO: if we own and stop GRTC, implement stop here */
```
Note: I don't think anybody is meant to ever stop the GRTC

aescolar reviewed (approved):
Approving understanding that this PR will be followed up shortly with another one fixing the entropy source so we avoid randomness in CI.

Thalley reviewed (approved):
It's passing existing tests, but are we testing all the new Kconfigs?
